tensor2tensor/utils/largebatch_optimizer.py

```diff
@@ -0,0 +1,163 @@
+    self._create_non_slot_variable(initial_value=0,
+                                   name="iter",
+                                   colocate_with=first_var)
+
+    # Create slots for the first and second moments and grad accumulator.
+    for v in var_list:
+      self._zeros_slot(v, "m", self._name)
+      self._zeros_slot(v, "v", self._name)
+      self._zeros_slot(v, "grad_acc", self._name)
+
+  def _get_iter_variable(self):
+    if context.executing_eagerly():
+      graph = None
+    else:
+      graph = ops.get_default_graph()
+    return self._get_non_slot_variable("iter", graph=graph)
+
+  def _prepare(self):
+    super(LargebatchAdamOptimizer, self)._prepare()
+    self._n_t = ops.convert_to_tensor(self._n, name="n")
+
+  def _apply_cond(self, apply_fn, grad, var, *args,**kwargs):
+    """Call `apply_fn only if the current counter value (iter) is zero. This
+    method couples common functionality for all _apply_*() implementations
+    in Adam.
+
+    Args:
+      apply_fn: Callback function for applying gradients.
+      grad: Gradients (grad variable in _apply_*() methods)
+      var: variable (var variable in _apply_*() methods)
+      *args: Passed through to `apply_fn`
+      **kwargs: Passed through to `apply_fn`
+    
+    Returns:
+      Adam op for applying gradients if iter=0, or an op for accumulating the
+      gradient in the `grad_acc` slot otherwise.
+    """
+    grad_acc = self.get_slot(var, "grad_acc")
+
+    def apply_adam(grad_acc, apply_fn, grad, var, *args, **kwargs):
+      total_grad = (grad_acc + grad) / math_ops.cast(self._n_t, 
+                                                     grad.dtype.base_dtype)
+      adam_op = apply_fn(total_grad, var, *args, **kwargs)
+      with ops.control_dependencies([adam_op]):
+        grad_acc_to_zero_op = grad_acc.assign(grad_acc * 0.0,
+                                              use_locking=self._use_locking)
+      return control_flow_ops.group(adam_op, grad_acc_to_zero_op)
+
+    def accumulate_gradient(grad_acc, grad):
+      assign_op = tf.assign_add(grad_acc, grad, use_locking=self._use_locking)
+      return control_flow_ops.group(assign_op)  # Strip return value
+      
+    return tf.cond(tf.equal(self._get_iter_variable(), 0),
+                   lambda: apply_adam(
+                     grad_acc, apply_fn, grad, var, *args, **kwargs),
+                   lambda: accumulate_gradient(grad_acc, grad))
+
+  def _apply_dense(self, grad, var):
+    return self._apply_cond(
+      super(LargebatchAdamOptimizer, self)._apply_dense, grad, var)
+
+  def _resource_apply_dense(self, grad, var):
+    return self._apply_cond(
+      super(LargebatchAdamOptimizer, self)._resource_apply_dense, grad, var)
+
+  def _apply_sparse_shared(self, grad, var, indices, scatter_add):
+    return self._apply_cond(
+      super(LargebatchAdamOptimizer, self)._apply_sparse_shared, grad, var,
+      indices, scatter_add)
+
+  def _apply_sparse(self, grad, var):
```

Comment:
Sure that I can do that? The optimizer works even for sparse tensors, just not as efficient as it could be as I'm simply converting to dense and use _apply_dense